### PR TITLE
Filtermedia process

### LINKED ISSUE #8062: Make filter-media available from the REST APi
**Is your feature request related to a problem? Please describe.**
Filter-media can currently only be triggered from the CLI. It would be good to convert it to a Scripts & Process script, so it can be triggered by admins from the UI as well

This can be done in an estimated **6** hours

### PR DESCRIPTION
**Pull request by YanaDePauw**
## References
Fixes #8062

## Description
This PR updates the existing filtermedia script to the version that can be used from both the CLI and the REST API. 

## Instructions for Reviewers
* Run the filtermedia from the CLI with the different parameter options. 
* Run the filtermedia from the REST API with the different parameter options. 

The script supports the following options:
```
 -f,--force              force all bitstreams to be processed
 -h,--help               help
 -i,--identifier <arg>   ONLY process bitstreams belonging to identifier
 -m,--maximum <arg>      process no more than maximum items
 -p,--plugins <arg>      ONLY run the specified Media Filter plugin(s)
                         listed from 'filter.plugins' in dspace.cfg.
                         Separate multiple with a comma (,)
                         (e.g. MediaFilterManager -p
                         "Word Text Extractor","PDF Text Extractor")
 -q,--quiet              do not print anything except in the event of
                         errors.
 -s,--skip <arg>         SKIP the bitstreams belonging to identifier
                         Separate multiple identifiers with a comma (,)
                         (e.g. MediaFilterManager -s
                         123456789/34,123456789/323)
 -v,--verbose            print all extracted text and other details to
                         STDOUT
```

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/app/mediafilter/MediaFilterScriptConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import java.sql.SQLException;
+
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.dspace.authorize.service.AuthorizeService;
+import org.dspace.core.Context;
+import org.dspace.scripts.configuration.ScriptConfiguration;
+import org.springframework.beans.factory.annotation.Autowired;
+
+public class MediaFilterScriptConfiguration<T extends MediaFilterScript> extends ScriptConfiguration<T> {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/mediafilter/MediaFilterScriptConfiguration.java:19]: Please add some basic JavaDocs to this new Class

**Review by tdonohue — CHANGES_REQUESTED** [review]
@YanaDePauw : This looks good overall, and it seems to work from the Commandline (at least from the options that I tried).

However, I cannot get this script to run from the UI's "/processes" page if I give it *no parameters* (which is the way this script is usually run).  Instead, I get an error:
```
core.js:6210 ERROR TypeError: Cannot read properties of undefined (reading 'map')
    at e.isRequiredMissing (process-form.component.ts:132)
    at e.submitForm (process-form.component.ts:72)
    at process-form.component.html:7
    at kl (core.js:15279)
    at Object.s [as next] (core.js:15317)
    at u.__tryOrUnsub (Subscriber.js:183)
    at u.next (Subscriber.js:122)
    at l._next (Subscriber.js:72)
    at l.next (Subscriber.js:49)
    at vg.next (Subject.js:39)
```

It appears (to me) that our `/processes` page may not currently support running a script with zero parameters?  

Also, if I run this from the UI with the `--force` parameter, the process output is always empty.  It appears the results of the script only go to `dspace.log`, as the Process Output ends up just saying this:
```
2021-12-20 17:09:31.324 INFO filter-media - 10 @ The script has started
2021-12-20 17:09:39.888 INFO filter-media - 10 @ The script has completed
```
I'd expect to see the same sort of information that you normally get to stdout when running this from the commandline.

Overall, I like the ideas here, I just think it's not working quite as expected yet.

**Comment by tdonohue** [comment]
@benbosman noted in today's meeting that the bug I ran into when running this from the UI appears to be logged in https://github.com/DSpace/dspace-angular/issues/1445  As noted today, this seems to imply that this PR may be semi-dependent on fixing that bug.

@artlowel : Any chance you or someone on your team could work on https://github.com/DSpace/dspace-angular/issues/1445?  I've tentatively assigned it to you as it sounds/looks small, so I'm hoping it could be easy to fit into 7.2

**Comment by mspalti** [comment]
@YanaDePauw , as expected I'm seeing the same angular error report.

The script also runs fine for me from the command line.

When I execute the process from the UI with a --force parameter it seems to execute fine, but I am seeing an error in the DSpace log. This happens after the `RestDSpaceRunnableHandler` logs the "The script has completed" message. The error is thrown inside a dispatcher that is trying to determine the parent item via the bundle. The bundle list is empty for the bitstream.

`java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64) ~[?:?]
	at jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70) ~[?:?]
	at jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248) ~[?:?]
	at java.util.Objects.checkIndex(Objects.java:372) ~[?:?]
	at java.util.ArrayList.get(ArrayList.java:459) ~[?:?]
	at org.hibernate.collection.internal.PersistentBag.get(PersistentBag.java:562) ~[hibernate-core-5.4.10.Final.jar:5.4.10.Final]
	at org.dspace.iiif.IIIFCacheEventConsumer.consume(IIIFCacheEventConsumer.java:79) ~[dspace-api-7.2-SNAPSHOT.jar:7.2-SNAPSHOT]`

I haven't determined what's causing this. It may not be important and it may be the dispatcher needs to be updated.

**Comment by mspalti** [comment]
The dispatcher error happens when the bitstream for the process log is created. This suggests to me that the problem is not here. The `IIIFCacheEventConsumer` may need to be updated to return when the bitstream has no associated bundle.  @tdonohue , @YanaDePauw , I can look into this you agree.

**Comment by YanaDePauw** [comment]
@mspalti I don't see an issue with fixing this. 

I took a quick look myself and the bitstream causing the issue is the output file of the script which does not have a parent object. Therefore it should be sufficient to just check if the bitstream has bundles before trying to get the first one.

**Review by tdonohue — COMMENTED** [review]
@YanaDePauw : I retested this today. It now works from the Angular UI (as https://github.com/DSpace/dspace-angular/issues/1445 was fixed)!

However, I'm still finding it much less useful from the Angular UI, as it only logs a start/stop date/time in the process logs like this:
```
2022-01-14 20:12:40.926 INFO filter-media - 14 @ The script has started
2022-01-14 20:12:43.155 INFO filter-media - 14 @ The script has completed
```

Any successful filtering & errors are only sent to `System.out` still.  This seems to be because **most of the filter-media logs** are generated from `MediaFilterServiceImpl`, which seems to primarily still use `System.out`, e.g. https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/mediafilter/MediaFilterServiceImpl.java#L343

All these `System.out` messages are lost when running this process from the UI.  So, you are not able to see which bitstreams were skipped, which were processed successfully, or even which ones threw errors.  The `dspace.log` only shows high level information, like a bitstream was updated or an error occurred, but there's no context around what was going on in the system at that time.

So, overall, while this PR "works", it's not very useful when running FilterMedia from the Admin UI, as you really get no information back at all -- this makes it extremely hard to determine what succeeded and what failed.   

Could you look at a possible strategy to ensure this information isn't lost from the UI?  If there's no way to achieve this easily, I'd be OK with possibly including this PR as-is, but I feel most users will be dissatisfied with the results.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @YanaDePauw !  This looks good to me now.  The logs in the Admin UI are now much more useful, and I've verified that the CLI output also still looks good.

@mspalti : I'll wait to merge this until you have a chance to re-verify, but it sounds like you were mostly in approval of this?  Please let me know so that I know whether it's ready to merge or not.

**Review by mspalti — APPROVED** [review]
I don't have time to re-verify today but my earlier test showed this worked with the exception of logging and the angular issues that have been noted and  fixed. I can approve now if others have already verified those fixes.

**Comment by tdonohue** [comment]
@mspalti : Yes, to verify, I've verified (by retesting) that the prior feedback has been addressed.

Therefore, since this is at +2, I'll merge this.  Thanks @YanaDePauw !

**Pull request merged**